### PR TITLE
fix: remove checking caret & dollar, add test

### PR DESCRIPTION
**Pull request by ivan-timchur-oat**
## Ticket
https://oat-sa.atlassian.net/browse/TR-3373

## Summary
What is it about?
ExtendedText or InlineText with pattern: submit fails if validateResponses is set

## How to test
1. Create item with constraint (pattern) or max words or max length
2. In test set "Validate Response"
3. Check test is working correct (in wrong variant it gives 500)

## Dependencies
No dependencies

## Sandbox environment
TODO

### CODE REVIEW

File: src/qtism/runtime/expressions/operators/Utils.php [hunk 1]
```diff
@@ -335,12 +335,32 @@ public static function prepareXsdPatternForPcre($pattern)
         // Neither caret (^) nor dollar ($) sign have special meaning so they are
         // considered as normal characters.
         // see http://www.regular-expressions.info/xml.html
-        $pattern = self::escapeSymbols($pattern, ['$', '^']);
-        $pattern = self::pregAddDelimiter('^' . $pattern . '$');
+//        $pattern = self::escapeSymbols($pattern, ['$', '^']);
```
[the remark this answers is not in the document]
Comment (ivan-timchur-oat) [reply src/qtism/runtime/expressions/operators/Utils.php:338]: also have same opinion, in slack write about this situation

File: src/qtism/runtime/expressions/operators/Utils.php [hunk 2]
```diff
@@ -335,12 +335,32 @@ public static function prepareXsdPatternForPcre($pattern)
...
+        $pattern = self::endDollar($pattern);
+        $pattern = self::pregAddDelimiter($pattern);
 
         // XSD regexp always case-sensitive (nothing to do), dot matches white-spaces (use PCRE_DOTALL).
         $pattern .= 's';
 
         return $pattern;
     }
+
+    private static function beginsCaret(string $pattern): string
```
[the remark this answers is not in the document]
Comment (ivan-timchur-oat) [reply src/qtism/runtime/expressions/operators/Utils.php:349]: it breaks the tests
```
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'/^(?:(?:[^\s\:\!\?\;\…\€]+)[\s\:\!\?\;\…\€]*){0,5}$/s'
+'/^\^(?:(?:[^\s\:\!\?\;\…\€]+)[\s\:\!\?\;\…\€]*){0,5}\$$/s'
```

[the remark this answers is not in the document]
Comment (wazelin) [reply src/qtism/runtime/expressions/operators/Utils.php:349]: You can modify `pregAddDelimiter()` implementation into something like this then.

```php
    public static function pregAddDelimiter(string $string): string
    {
        return '/' . static::escapeSymbols($string, '/') . '/';
    }

    public static function trimPregDelimiter(string $string): string
    {
        if (strlen($string) < 3 || $string[0] !== $string[-1]) {
            return $string;
        }

        return trim($string, $string[0]);
    }
```

File: src/qtism/runtime/expressions/operators/Utils.php [hunk 3]
```diff
@@ -188,14 +191,14 @@ public static function standardDeviation(array $sample, $correction = true)
      */
     public static function pregAddDelimiter($string)
     {
-        return '/' . static::escapeSymbols($string, '/') . '/';
+        return '/'.static::escapeSymbols($string, '/').'/';
```
Comment (quintanilhar) [inline src/qtism/runtime/expressions/operators/Utils.php:194]: Missing spaces

**Review by quintanilhar — APPROVED** [review]
All good! :)

- [X] New code is covered by tests
- [X] Tests are running successfully on my local machine
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X] Feature is working correctly on my local machine
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful
- [X] Pull request's target is not `master`
- [X] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [X] Commits messages are meaningful
- [X] Commits are atomic
-- - [ ] Changelog is updated according to changes --
-- - [ ] Documentation is updated according to changes --

**Review by quintanilhar — APPROVED** [review]
All good! :)

- [X] New code is covered by tests
- [X] Tests are running successfully on my local machine
- [X] New code is respecting code style rules
- [X] New code is respecting best practices
- [X] Feature is working correctly on my local machine
- [X] Acceptance criteria are respected
- [X] Pull request title and description are meaningful
- [X] Pull request's target is not `master`
- [X] Commits are following [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [X] Commits messages are meaningful
- [X] Commits are atomic
- [x] Changelog is updated according to changes
- [ ] Documentation is updated according to changes

**Comment by julien-sebire** [comment]
I'm not sure you should remove the escaping at all.
The idea is that the pattern you receive from the item definition is not PCRE but XML regex, in which the `$` and `^` are not anchors but just regular characters for dollar and caret, and the whole string is searched. For example, if you want the assert a price, you can write something like `$[0-9]+`. The $ here is meant to be a monetary symbol, not a regex anchor. So the escaping here is very much wanted so that the $ is not mistakenly seen as an anchor when you inject this pattern to `preg_match`. 
What should be changed is what is sent by the Front End in the first place, to include only the XML regex. If the Back End receives a $ sign, it's a $ sign we want to find in the string, not an anchor. And considering the XML should always match the whole string, then this is why we add the anchors to feed `preg_match`.
And before that, we shoudl fix the Authoring Front End to fix the persistence of the regex in the first place, which must not have anchors.

**Review by julien-sebire — CHANGES_REQUESTED** [review]
:warning: I'm not sure you should remove the escaping at all.

The idea is that the pattern you receive from the item definition is not PCRE but XML regex, in which the `$` and `^` are not anchors but just regular characters for dollar and caret, and the whole string is searched. For example, if you want to assert a price in dollars, you can write something like `$[0-9]+`. The `$` here is meant to be a monetary symbol, not a regex anchor. So the escaping here is very much wanted so that the `$` is not mistakenly seen as an anchor when you inject this pattern to `preg_match`. But it is then very difficult to tell if we expect to receive a PCRE or an XML regex.

What should be changed is what is sent by the Front End in the first place, to include only the XML regex. Then, no ambiguity, if the Back End receives a `$` sign, it's a `$` sign we want to find in the string, not an anchor. And considering the XML should always match the whole string, then this is why we add the anchors to feed `preg_match`.

And before that, we should fix the Authoring Front End to fix the persistence of the regex in the first place, which is currently saving PCREs instead of XML regexes (must not have anchors).

In short, this PR should be rejected, and one created in the Item Runner Front End to send an XML regex, and another one in the Autoring tool Front End.

**Comment by ivan-timchur-oat** [comment]
> $[0-9]+
add such variant to test:
'dollar with number' => ['$[0-9]+', '/^\$[0-9]+$/s',],

php vendor/phpunit/phpunit/phpunit test/qtismtest/runtime/expressions/operators/OperatorsUtilsTes
t.php
PHPUnit 9.5.13 by Sebastian Bergmann and contributors.

Warning:       Your XML configuration validates against a deprecated schema.
Suggestion:    Migrate your XML configuration using "--migrate-configuration"!

................................................................. 65 / 78 ( 83%)
.............                                                     78 / 78 (100%)

Time: 00:00.497, Memory: 8.00 MB


^\$[0-9]+$ the test shows, that we will escape the dollar sign, of course there can be very specific patterns, but current realization is most soft.

**Comment by quintanilhar** [comment]
> The idea is that the pattern you receive from the item definition is not PCRE but XML regex, in which the $ and ^ are not anchors but just regular characters for dollar and caret, and the whole string is searched. For example, if you want to assert a price in dollars, you can write something like $[0-9]+. The $ here is meant to be a monetary symbol, not a regex anchor. So the escaping here is very much wanted so that the $ is not mistakenly seen as an anchor when you inject this pattern to preg_match. But it is then very difficult to tell if we expect to receive a PCRE or an XML regex.

The changes don't remove the escaping logic, just avoid escaping the anchors which were used improperly. For example:
```
'max chars string pattern without anchors' => ['[\s\S]{0,5}', '/^[\s\S]{0,5}$/s'],
'max chars string pattern with anchors' => ['^[\s\S]{0,5}$', '/^[\s\S]{0,5}$/s'],
```

Both versions should generate the same regex IMHO. As mentioned before, in Terre, the generated attribute `patternMatch` is including the anchors, therefore, breaking the application. The solution created by @ivan-timchur-oat prevents that as you can see in the test cases.

> What should be changed is what is sent by the Front End in the first place, to include only the XML regex. Then, no ambiguity, if the Back End receives a $ sign, it's a $ sign we want to find in the string, not an anchor. And considering the XML should always match the whole string, then this is why we add the anchors to feed preg_match.

I agree, but we also need to consider all the faulty test packages generated with the anchors. I don't think this solution we'll generate issues in a long term.

> And before that, we should fix the Authoring Front End to fix the persistence of the regex in the first place, which is currently saving PCREs instead of XML regexes (must not have anchors).

I'd really appreciate it if you can create another bug ticket to address this issue.

> In short, this PR should be rejected, and one created in the Item Runner Front End to send an XML regex, and another one in the Autoring tool Front End.

We were considering that in the first place but I still believe that having a solution that is future-proof will bring benefits for everyone and also support any previous test package created with this wrong regular expression. My call is to proceed.

File: src/qtism/runtime/expressions/operators/Utils.php [hunk 4]
```diff
@@ -82,7 +82,8 @@ public static function lcm($a, $b)
      * Compute the arithmetic mean of $sample.
      *
      * @param array An array of numeric values.
-     * @return false|number The arithmetic mean of $sample or false if any of the values of $sample is not numeric or if $sample is empty.
+     * @return false|number The arithmetic mean of $sample or false if any of the values of $sample is not numeric or
+     *     if $sample is empty.
```
Comment (poyuki) [inline src/qtism/runtime/expressions/operators/Utils.php:86]: are this line break necessary?

**Review by wazelin — APPROVED** [review]
- [x] New code is covered by tests (if applicable)
- [x] Tests are running successfully (old and new ones) on my local machine (if applicable)
- [x] New code is respecting code style rules
- [x] New code is respecting best practices
- [x] New code is not subject to concurrency issues (if applicable)
- [x] Feature is working correctly on my local machine (if applicable)
- [x] Acceptance criteria are respected
- [x] Pull request title and description are meaningful
- [x] Pull request's target is not `master`
